### DISCUSSION TITLE
Nanite-style DAG clusterizing and simplification

### DISCUSSION THREAD

**Discussion by JMS55** [zeux/meshoptimizer#750] [General]
Followup to https://github.com/bevyengine/bevy/issues/14660

In Bevy's Nanite-style renderer, I've run into a couple problems:
* Can't get down to 1 root cluster (or even close to it) - The grouping/simplification stalls out at a certain point, and we're left unable to make coarser LODs
* Poor meshlet occupancy and irregular/tiny meshlets - I simplify to 64/64 triangles/vertices max, but most meshlets are well under, leaving workgroups half empty. Similarly, there's a bunch of tiny meshlets with very few triangles. 

Nanite was obviously able to get to much better results, and there are 3 major parts:
* Grouping triangles into meshlets
* Grouping meshlets together in preparation for simplification
* Simplifying groups of merged meshlets 

For the first two, I use meshoptimizer. The latter I do myself with METIS. 

I'm opening this thread to hopefully improve my mesh -> meshlet mesh building code. Maybe meshoptimizer can improve in some way, and my code almost certainly so.

**Comment by JMS55** [comment]
First step is to collect some hard data. I need to run a couple of different meshes through the converter and collect some statistics on their DAG levels, meshlets per level, meshlet occupancy, etc.

I also need to try the new attribute-aware simplification on meshoptimizer main.

**Reply by zeux** [reply]
As an aside, is the converter part of Bevy? I wasn't able to continue testing some of the changes in Bevy because I've been using an older glTF-based PR that became unmergeable. If the converter was buildable and runnable easily that could make it simpler to validate some of the ideas.

**Reply by JMS55** [reply]
Not easily, unfortunately. The current workflow is to open a gltf in blender, merge everything down to 1 mesh, and then run it through a janky program I didn't merge upstream. More ergonomic and built-in workflows are blocked on larger design changes in Bevy - the gltf converter I had was not a good path.

I will try to cleanup the program and upload it here to make testing easier.

**Reply by zeux** [reply]
Got it. I think it will be good long term for meshopt to have a workflow demo that approximates Nanite so I will probably look into that next week, so may not need this.

**Reply by JMS55** [reply]
Collected some data on the stanford bunny, stanford dragon, arcane jinx from sketchfab (merged down to 1 mesh), and quixel megascan's "Huge Icelandic Lava Cliff". Meshlet occupancy is a map of `triangle_count:meshlet_count`, i.e. the number of meshlets with that triangle count per LOD.

```
bunny.glb
LOD: 0, meshlet count: 2251, meshlet occupancy counts: {64: 2250, 42: 1}
LOD: 1, meshlet count: 1320, meshlet occupancy counts: {64: 931, 20: 1, 61: 1, 32: 384, 6: 1, 58: 1, 3: 1}
LOD: 2, meshlet count: 672, meshlet occupancy counts: {42: 1, 61: 1, 48: 135, 59: 3, 47: 1, 54: 1, 4: 1, 58: 1, 38: 1, 34: 2, 16: 26, 49: 5, 62: 4, 33: 1, 3: 1, 44: 1, 64: 383, 21: 2, 32: 82, 2: 3, 39: 1, 17: 1, 63: 6, 5: 1, 55: 1, 60: 2, 57: 1, 1: 1, 56: 2, 40: 1}
LOD: 3, meshlet count: 373, meshlet occupancy counts: {26: 2, 61: 8, 8: 16, 51: 1, 25: 1, 48: 10, 59: 1, 47: 1, 41: 1, 54: 2, 4: 6, 23: 1, 58: 6, 52: 2, 34: 4, 16: 15, 62: 4, 28: 1, 11: 2, 35: 2, 27: 1, 13: 1, 3: 2, 29: 1, 64: 172, 21: 2, 9: 3, 32: 17, 44: 1, 24: 19, 17: 2, 63: 7, 12: 2, 20: 1, 7: 1, 5: 1, 19: 2, 2: 1, 50: 2, 60: 10, 36: 1, 22: 2, 57: 3, 1: 2, 56: 23, 40: 6, 31: 2}
LOD: 4, meshlet count: 173, meshlet occupancy counts: {43: 1, 61: 8, 42: 1, 25: 3, 51: 2, 8: 2, 47: 1, 46: 2, 59: 6, 54: 1, 4: 1, 41: 1, 58: 5, 30: 2, 23: 1, 38: 1, 37: 1, 34: 3, 62: 10, 49: 1, 16: 1, 11: 1, 13: 2, 3: 3, 44: 2, 64: 47, 21: 2, 29: 2, 32: 2, 2: 1, 24: 3, 63: 10, 12: 3, 17: 1, 20: 3, 55: 3, 7: 1, 53: 2, 10: 1, 50: 3, 9: 1, 60: 9, 45: 1, 36: 1, 57: 4, 22: 2, 26: 3, 56: 2, 40: 4}
LOD: 5, meshlet count: 74, meshlet occupancy counts: {43: 1, 42: 1, 8: 1, 61: 3, 25: 1, 41: 2, 59: 3, 4: 2, 30: 1, 58: 2, 62: 3, 11: 1, 3: 1, 64: 15, 21: 2, 32: 2, 39: 1, 19: 2, 63: 6, 14: 2, 10: 1, 55: 1, 6: 2, 53: 2, 50: 1, 15: 1, 60: 4, 57: 3, 1: 1, 26: 1, 56: 4, 40: 1}
LOD: 6, meshlet count: 19, meshlet occupancy counts: {29: 1, 64: 4, 61: 2, 28: 1, 62: 1, 60: 1, 49: 1, 33: 1, 63: 1, 54: 1, 55: 1, 53: 2, 7: 1, 50: 1}

dragon.glb
LOD: 0, meshlet count: 1575, meshlet occupancy counts: {55: 22, 25: 4, 62: 16, 42: 18, 51: 16, 60: 12, 61: 15, 41: 13, 44: 18, 63: 15, 56: 14, 46: 8, 64: 1064, 38: 20, 39: 6, 34: 21, 49: 13, 57: 16, 52: 9, 37: 16, 40: 20, 45: 14, 30: 14, 29: 3, 26: 5, 43: 13, 48: 10, 32: 17, 33: 14, 59: 12, 36: 15, 53: 11, 54: 13, 31: 5, 50: 13, 28: 8, 47: 17, 58: 10, 27: 7, 35: 18}
LOD: 1, meshlet count: 870, meshlet occupancy counts: {55: 11, 25: 3, 62: 21, 19: 3, 42: 6, 10: 6, 22: 1, 23: 3, 41: 7, 56: 15, 38: 13, 46: 10, 34: 11, 1: 8, 49: 19, 40: 8, 57: 6, 37: 5, 16: 2, 45: 14, 21: 4, 48: 13, 18: 5, 33: 12, 43: 6, 17: 3, 54: 8, 14: 3, 20: 2, 50: 15, 6: 10, 28: 8, 58: 22, 27: 4, 35: 7, 8: 6, 11: 4, 51: 13, 60: 25, 61: 24, 7: 4, 44: 4, 63: 18, 64: 260, 30: 6, 39: 12, 5: 7, 52: 15, 29: 4, 12: 3, 32: 70, 4: 11, 9: 6, 15: 4, 59: 15, 36: 11, 53: 9, 13: 5, 31: 9, 24: 5, 3: 10, 2: 7, 47: 9}
LOD: 2, meshlet count: 145, meshlet occupancy counts: {55: 2, 25: 2, 62: 9, 11: 1, 42: 1, 51: 2, 22: 1, 23: 1, 60: 8, 61: 9, 7: 2, 44: 4, 63: 7, 56: 4, 64: 20, 46: 2, 39: 1, 30: 2, 5: 5, 52: 4, 16: 1, 29: 1, 57: 5, 40: 1, 10: 1, 45: 1, 21: 2, 32: 6, 18: 1, 43: 3, 33: 1, 9: 2, 59: 6, 15: 1, 48: 1, 36: 1, 53: 3, 54: 1, 17: 1, 14: 1, 20: 1, 6: 2, 3: 4, 2: 2, 47: 1, 58: 6, 8: 2}
LOD: 3, meshlet count: 5, meshlet occupancy counts: {41: 1, 52: 1, 63: 1, 5: 1, 64: 1}

cliff.glb
LOD: 0, meshlet count: 31244, meshlet occupancy counts: {59: 12, 57: 4, 45: 2, 61: 13, 48: 3, 55: 2, 64: 31118, 56: 12, 50: 1, 52: 5, 63: 25, 54: 6, 60: 8, 49: 3, 32: 1, 62: 12, 51: 1, 53: 5, 33: 1, 28: 1, 58: 9}
LOD: 1, meshlet count: 18265, meshlet occupancy counts: {36: 7, 12: 1, 61: 14, 55: 3, 6: 1, 34: 9, 39: 3, 7: 2, 64: 12908, 23: 1, 50: 2, 52: 2, 32: 5069, 28: 11, 54: 3, 60: 18, 14: 2, 62: 21, 2: 9, 51: 3, 41: 2, 53: 2, 38: 2, 47: 1, 18: 1, 1: 15, 3: 5, 10: 1, 57: 6, 59: 11, 49: 1, 37: 7, 48: 3, 40: 3, 43: 1, 24: 4, 30: 14, 26: 6, 25: 2, 56: 4, 5: 4, 4: 6, 63: 20, 33: 5, 44: 2, 35: 8, 27: 5, 29: 5, 31: 22, 8: 2, 58: 5, 20: 1}
LOD: 2, meshlet count: 9811, meshlet occupancy counts: {36: 4, 12: 7, 61: 33, 39: 3, 7: 4, 23: 5, 15: 14, 60: 29, 52: 3, 54: 8, 62: 41, 2: 18, 21: 7, 18: 15, 10: 8, 48: 809, 40: 3, 24: 4, 13: 4, 30: 10, 26: 2, 25: 2, 35: 4, 29: 3, 11: 1, 31: 9, 19: 9, 9: 5, 42: 3, 6: 8, 55: 12, 34: 7, 43: 2, 64: 5997, 50: 12, 32: 970, 28: 4, 14: 29, 49: 9, 41: 3, 51: 9, 53: 7, 17: 15, 38: 3, 47: 6, 22: 10, 1: 40, 59: 19, 57: 14, 3: 22, 45: 3, 16: 1381, 56: 16, 4: 9, 5: 9, 63: 69, 46: 14, 33: 8, 20: 10, 44: 3, 27: 2, 8: 5, 58: 17}
LOD: 3, meshlet count: 5122, meshlet occupancy counts: {36: 8, 12: 19, 61: 84, 39: 10, 7: 20, 15: 19, 23: 11, 60: 71, 52: 17, 54: 35, 62: 106, 2: 23, 21: 8, 18: 18, 10: 18, 48: 184, 37: 9, 40: 203, 24: 169, 13: 13, 26: 10, 30: 17, 25: 13, 35: 11, 29: 9, 11: 15, 31: 14, 19: 12, 9: 31, 42: 14, 6: 22, 55: 26, 34: 10, 43: 10, 64: 2385, 50: 14, 32: 244, 41: 13, 14: 20, 49: 7, 28: 8, 51: 11, 53: 19, 17: 22, 38: 13, 47: 16, 1: 38, 3: 25, 59: 57, 57: 33, 22: 13, 45: 8, 16: 209, 56: 219, 4: 27, 5: 19, 63: 102, 33: 15, 46: 10, 20: 14, 27: 13, 44: 7, 8: 233, 58: 49}
LOD: 4, meshlet count: 2547, meshlet occupancy counts: {36: 24, 12: 24, 61: 105, 39: 11, 7: 18, 15: 21, 23: 8, 60: 111, 52: 29, 54: 28, 62: 97, 2: 19, 21: 14, 18: 13, 10: 25, 37: 12, 48: 28, 40: 41, 13: 24, 24: 18, 26: 15, 30: 15, 25: 9, 29: 8, 35: 16, 11: 25, 31: 19, 19: 15, 42: 6, 9: 19, 55: 32, 6: 19, 34: 10, 43: 11, 64: 798, 50: 16, 32: 47, 28: 36, 41: 14, 14: 17, 49: 11, 51: 12, 53: 23, 17: 12, 38: 10, 47: 5, 1: 29, 22: 16, 59: 80, 57: 38, 3: 22, 45: 5, 16: 28, 56: 65, 4: 38, 5: 30, 63: 103, 46: 15, 33: 11, 44: 30, 27: 17, 20: 26, 8: 36, 58: 68}
LOD: 5, meshlet count: 945, meshlet occupancy counts: {36: 7, 12: 15, 61: 57, 7: 3, 39: 10, 23: 9, 15: 4, 60: 64, 52: 13, 54: 18, 2: 8, 62: 53, 21: 6, 18: 6, 10: 10, 48: 4, 37: 4, 40: 12, 13: 9, 24: 8, 26: 7, 30: 5, 25: 6, 35: 11, 29: 2, 11: 6, 31: 5, 19: 7, 9: 6, 42: 5, 55: 23, 6: 8, 34: 3, 43: 7, 64: 149, 50: 9, 32: 10, 49: 9, 41: 9, 14: 5, 28: 4, 51: 5, 53: 23, 17: 4, 38: 7, 47: 11, 22: 2, 1: 4, 59: 45, 57: 45, 45: 8, 3: 6, 16: 7, 56: 17, 5: 3, 4: 4, 63: 44, 33: 9, 46: 8, 20: 6, 27: 8, 44: 5, 8: 6, 58: 52}
LOD: 6, meshlet count: 221, meshlet occupancy counts: {36: 1, 9: 1, 61: 17, 12: 1, 55: 8, 6: 1, 34: 2, 7: 4, 39: 1, 64: 25, 50: 2, 60: 10, 14: 1, 41: 3, 32: 4, 49: 4, 28: 2, 62: 17, 51: 1, 53: 2, 21: 3, 2: 1, 54: 5, 18: 1, 52: 1, 1: 3, 22: 1, 59: 12, 57: 9, 45: 1, 3: 2, 37: 2, 48: 1, 40: 1, 16: 2, 13: 1, 24: 2, 30: 3, 26: 1, 25: 3, 56: 6, 5: 1, 4: 1, 63: 14, 33: 2, 46: 3, 44: 1, 27: 4, 29: 2, 11: 1, 8: 6, 31: 1, 58: 15, 19: 2}
LOD: 7, meshlet count: 16, meshlet occupancy counts: {60: 2, 59: 1, 63: 2, 64: 1, 43: 1, 48: 1, 56: 1, 62: 1, 54: 1, 9: 1, 26: 1, 39: 1, 25: 1, 47: 1}

jinx.glb
LOD: 0, meshlet count: 103, meshlet occupancy counts: {63: 2, 4: 1, 52: 1, 61: 1, 60: 2, 64: 95, 57: 1}
LOD: 1, meshlet count: 64, meshlet occupancy counts: {63: 1, 54: 1, 49: 1, 64: 34, 36: 1, 35: 1, 59: 2, 3: 2, 34: 1, 61: 2, 58: 1, 32: 13, 9: 1, 51: 1, 8: 1, 39: 1}
LOD: 2, meshlet count: 27, meshlet occupancy counts: {63: 1, 61: 1, 10: 1, 64: 10, 32: 2, 36: 1, 30: 1, 48: 3, 52: 1, 43: 1, 60: 1, 2: 1, 17: 1, 59: 2}
LOD: 3, meshlet count: 5, meshlet occupancy counts: {9: 1, 64: 2, 1: 1, 59: 1}
```

**Reply by JMS55** [reply]
Cluster counts.... actually looks pretty good now? I guess something has changed since my initial implementation several months ago. Not able to get to a _single_ root cluster, but <20 is fairly good. We could probably do better once I tune the heuristics to be more aggressive.

Meshlet occupancy counts I haven't yet analyzed, need to build some graphs. But it looks like at least for LOD 0, we're doing pretty good.

**Reply by JMS55** [reply]
![image](https://github.com/user-attachments/assets/4b3ce3f9-6603-460b-b554-74f0ffd02096)
I skipped around a bit, but here's LODs 0,1,2,5 (missing 3, 4, 6, 7) for the cliffs mesh. 

At LOD 0, 31118/31244 ~= 99.6% of meshlets have 64 triangles
At LOD 1, 12908/18265 ~= 70.7% of meshlets have 64 triangles
At LOD 2, 5997/9811 ~= 61.1% of meshlets have 64 triangles
At LOD 5, 149/945 ~= 15.7% of meshlets have 64 triangles

At each further LOD level, the amount of full meshlets decreases.

Interestingly, the amount of meshlets with 48/32/16 triangles increases per LOD level proportionately.

**Comment by zeux** [comment]
Worth noting in general that you will never have fully balanced workgroups with 64v/64t: on grid-like meshes a 8x4 regular quad grid (8x4 * 2 = 64 triangles) uses 9*5=45 vertices; on the flip side, if you have a lot of disjointed triangles, you will have a lot of vertices but few triangles, e.g. worst case is 21 triangles with 3 unique vertices each => 63 vertices. That's fine in the sense that it's a common property of any clustered renderer to have an imbalance somewhere, just noting that if you expect 64v/64t then that's not going to happen very often. See https://zeux.io/2023/01/16/meshlet-size-tradeoffs/

But otherwise I think it's critical to understand *where* the issue happens and *why*; I would probably recommend starting with a reasonably simple mesh that should *not* have a problem with getting down to very few root clusters and trying to debug the steps that get stuck early. Could also try repeating that with Nanite in UE to understand what data it generates.

**Reply by JMS55** [reply]
I don't actually need a balanced workflow for vertices. I mainly care about triangle counts per meshlet. My rasterizer is a 64-thread compute shader where each thread loads a vertex, transforms, and stores in LDS. Then each thread loads 3 vertices from LDS and rasterizes a triangles. The triangle raster is much more expensive than the vertex transforming.

> But otherwise I think it's critical to understand where the issue happens and why; I would probably recommend starting with a reasonably simple mesh that should not have a problem with getting down to very few root clusters and trying to debug the steps that get stuck early. Could also try repeating that with Nanite in UE to understand what data it generates.

Agreed. I need to install UE and see if how good a job they do at simplifying the bunny (which seems like a simple mesh). No time today, but will hopefully be back with some hard data later this week.

**Reply by zeux** [reply]
Got it. One other thing you could try is see if using a 128v/64t or 196v/64t configuration helps at all. This is aggressive (normally you wouldn't want this many vertices), and it can reduce occupancy by requiring more LDS for vertex storage, but given a complete triangle soup where every triangle is not connected to any other triangle as an input, that's the configuration you would need to maintain fully loaded wavefronts; this is not necessarily something you'd want to keep long term but trying that would also help figure out the limits of the system.

(to actually get "a triangle soup input" to work in your setup you would need to use simplify_with_locks which was added so that the boundary of a cluster can be explicitly specified to the simplifier, but I don't think you actually need to do this at this stage if you are testing on well formed meshes, it's just an example of an input asset that maximizes the pressure on the vertex data)

**Comment by zeux** [comment]
One other note: currently, `meshopt_buildMeshlets` is using a greedy algorithm that sometimes missteps and ends up creating smaller sub-clusters that aren't directly attached to the cluster itself. While disconnected clusters can't be avoided in general (eg if the mesh itself is disconnected), this definitely happens more often than it's necessary. This is less of a problem for direct clustered rendering that this was intended for but is much more of an issue for a Nanite like hierarchical pipeline since any such split will propagate downstream and create issues at lower LODs.

I'd like to improve this in the future somehow but I'm not sure how :sweat_smile: meanwhile I'd note that Nanite uses METIS to recursively split clusters in half until reaching a size close to the max required. It should be simple to try that approach in Bevy; my long term goal is for METIS to be unnecessary in Nanite like pipelines but it's valuable to understand if that's the core part of the issue and to what extent this is better.

**Comment by JMS55** [comment]
@zeux some questions:
* Is there any reason simplify vertex count is limited to 255? Could you increase it to 256? I'd like to use 256v/128t, I'm getting much better results with 255v/128t than 64v/64t.
* Is it correct that: passing target_error = f32::MAX_VALUE to simplify should mean that meshoptimizer will simplify as much as possible in order to reach the target index (triangle) count?

**Reply by zeux** [reply]
- 255 is uses by the clusterizer internally as a sentinel; I don’t think you should see meaningfully different results with 255 vs 256?
-  Yeah.

**Reply by JMS55** [reply]
* That's fair. I don't think it would be meaningfully different yeah, but it would be a nice round power of 2 😅 
* Thanks for confirming! This gives me much better results FYI, no reason to limit the target error per LOD simplify when the cluster culling code will choose an appropriate LOD anyways.

**Reply by zeux** [reply]
Yeah it’s certainly a little awkward. Maybe this can be fixed without switching to a different data type internally; I’ll need to check.

**Comment by Scthe** [comment]
@JMS55 I did some more tests in a [separate repo](https://github.com/Scthe/nanite-simplification-tests/tree/master). The README has most of what I found interesting. Adding some extended info below.

**EDIT 06-09-2024:** Added stats the other border vertex definition (vertex shared between meshlets instead of based only on geometry). The model's I've choose greatly influence results.

---

Overall problems:

1. Nanite does not allow you to simplify vertices that are shared with other meshlets. This is usually 30-50% of the meshlet's vertices (see log files in the repo) when constructing the first DAG level. It gets worse on higher levels.
    * Border defined based on the geometry (edge that belongs to a single triangle). If border is defined as verts shared be meshlets it's <30% and goes down to 0% higher up the hierarchy.
1. At some point, you end up with disjointed triangles. Not possible to simplify. Per meshoptimizer readme: 

> For meshes with inconsistent topology or many seams, such as faceted meshes, it can result in simplifier getting "stuck" and not being able to simplify the mesh fully. Therefore it's critical that identical vertices are "welded" together, that is, the input vertex buffer does not contain duplicates. Additionally, it may be worthwhile to weld the vertices without taking into account vertex attributes that aren't critical and can be rebuilt later.

With these 2 issues, I don't think it's possible to create a good DAG. The only solution seems to be randomly removing triangles (!). **At higher DAG levels all meshlets are disconnected single triangles.**

In the [logs](https://github.com/Scthe/nanite-simplification-tests/tree/master/logs) directory I've added console dumps with stats. These are the same models as I've [originally wrote](https://github.com/bevyengine/bevy/discussions/14998#discussion-7116394) month ago - [Jinx](https://github.com/Scthe/nanite-webgpu/tree/master/static/models/jinx-combined) ( `logs/jinx-*.txt` files) and [Modular Mecha Doll Neon Mask](https://sketchfab.com/3d-models/modular-mecha-doll-neon-mask-1e0dcf3e016f4bc897d4b39819220732) ( `logs/robot-*.txt` files).

* [jinx.txt](https://github.com/Scthe/nanite-simplification-tests/blob/master/logs/jinx.txt) - **without** random triangle removal. Ends up as 36 root meshlets (3717 triangles).
* [jinx-rngTrisRemove.txt](https://github.com/Scthe/nanite-simplification-tests/blob/master/logs/jinx-rngTrisRemove.txt) - **with** random triangle removal. Ends up as 1 root meshlet with 128 triangles.
* [jinx-sharedBorder-rngTrisRemove.txt](https://github.com/Scthe/nanite-simplification-tests/blob/master/logs/jinx-sharedBorder-rngTrisRemove.txt) - **with** random triangle removal. Border defined as vertices shared between meshlets.
* [robot.txt](https://github.com/Scthe/nanite-simplification-tests/blob/master/logs/robot.txt) - the "Modular Mecha Doll Neon Mask" model **without** random triangle removal. Ends up as 1597 root meshlets (176614 triangles).
* [robot-rngTrisRemove.txt](https://github.com/Scthe/nanite-simplification-tests/blob/master/logs/robot-rngTrisRemove.txt) - the "Modular Mecha Doll Neon Mask" model **with** random triangle removal. Ends up as 1 root meshlets with 128 triangles.
* [robot-sharedBorder-rngTrisRemove.txt](https://github.com/Scthe/nanite-simplification-tests/blob/master/logs/robot-sharedBorder-rngTrisRemove.txt) - **with** random triangle removal. Border defined as vertices shared between meshlets.

Ofc. it might sound stupid. You get perfect DAG if you randomly remove triangles! But what are the alternatives?

PS. DO NOT LOOK INTO [quadric.ts](https://github.com/Scthe/nanite-simplification-tests/blob/master/src/meshPreprocessing/quadric.ts).

**Comment by JMS55** [comment]
@zeux an issue I've ran into: I keep running into `Assertion failed: index_count / 3 <= kMeshletMaxTriangles` when calling `meshopt_computeClusterBounds`. I'm trying to compute a bounding sphere for meshlet groups (e.g. 8 meshlets merged together), which is over the triangle limit the function assumes :)

Might be nice to have an override for that. I'm going to try and patch meshopt with a higher limit to allow me to test my code.

**Reply by zeux** [reply]
Yeah, that function is meant for use by meshlets atm; it does not account for the possibility of using a larger collection of triangles. The limit is documented.

The good news is that using this function is probably incorrect anyway. I think the logic that Bevy is using atm is distance to sphere center, but even if it uses a distance to sphere boundary, the resulting error function is not monotonic unless the bounds of the merged group entirely contain the bounds of the children. To get here, you need to actually union the spheres into a larger sphere, as some spheres may be outside of a precise spherical boundary created around the vertices.

I ran into this yesterday when working on https://github.com/zeux/meshoptimizer/pull/760 (absolutely no promises any of that code is correct!); I wrote a function that validates that error through DAG is monotonic and discovered it's not.

**Reply by JMS55** [reply]
Yeah, at runtime, Bevy should be moving the group sphere to be on the closest point of the culling sphere, and _then_ projecting the group sphere. In practice it hasn't mattered too much.

Same with the group spheres not necessarily perfectly bounding their children - it hasn't been an issue in practice yet.

**Reply by zeux** [reply]
I don't think the culling sphere comment is right. To have conservative error, you need to make sure to match the computation of the parent LOD on the child cluster and self LOD on the parent cluster exactly; since clusters have multiple parents, these all should all agree on the same error - but you don't want multiple clusters to all have the same culling sphere.

And yeah I don't think the problem is very prevalent. It happened on some meshes in some configurations I was testing but not others. What can happen as a result though I think is duplicate geometry being rendered (or maybe gaps? unsure).

Either way, I've added a note to see if this limit can be completely removed; that requires a memory allocation which this function doesn't do atm. I don't want to increase this limit as the stack space requirements here become non-trivial for Wasm in particular.

**Reply by JMS55** [reply]
Hmm yeah, I might've misremembered the issue. It was something about projecting the group sphere though.
